### PR TITLE
Sanitycheck: tweaks to allow --device-testing --build-only to work

### PR DESCRIPTION
**Pull request by galak**
This is a step towards having support in sanitycheck to produce just the build artifacts for tests that are runnable on a board.

### CODE REVIEW

**Comment by nashif** [comment]
> This is a step towards having support in sanitycheck to produce just the build artifacts for tests that are runnable on a board.

what would be the next step?

IMO getting a list of runnable tests and ust build them or execute them should be done independent of --device-testing. We would want this to work for simulation and emulation etc.

**Comment by galak** [comment]
> what would be the next step?

The workflow I'm thinking is this:

CI build:

1. `./scripts/sanitycheck -p frdm_k64f --build-only  --device-testing -M`

This would build and strip the sanity-out dir (via `-M`) down to just the artifacts you need to have the tests run.

2. on successful build - tar up artifacts:  `sanity-out/`, `scripts/sanity*`, `scripts/west*`, `west.yml`

CI run:

3. `./scripts/sanitycheck -p frdm_k64f --test-only  --device-testing -i --device-serial /dev/ttyACM0 --west-flash`
 
> IMO getting a list of runnable tests and ust build them or execute them should be done independent of --device-testing. We would want this to work for simulation and emulation etc.

Agreed, I'd like to see simulation/emulation testing work under the same flow as device-testing.

Would we support `./scripts/sanitycheck --device-testing -p qemu_x86` ?

**Comment by erwango** [comment]
I wonder about possible confusion that would get out of this. What would be the diff between 
```
sanitycheck  -b -p my_board 
```
and
```
sanitycheck --device-testing --build-only
```
?

When running the following:
```
sanitycheck -p disco_l475_iot1 -b -T tests/kernel/common/
```
and
```
sanitycheck -p disco_l475_iot1 --device-testing --hardware-map ../map.yaml -T tests/kernel/common/
```
the only difference I get in output artefacts are sanitycheck.csv, sanitycheck.log and sanitycheck.xml, which are only related to the pure execution (please correct me if I'm wrong).
So, I don't really get what the `--build-only` would add in that context.

**Comment by galak** [comment]
> I wonder about possible confusion that would get out of this.

Not following

> What would be the diff between `sanitycheck -b -p my_board ` and `sanitycheck --device-testing --build-only` ?

The difference is in the filter of which tests will be built.  Adding `--device-testing` will have the list of tests only include those that are runnable on the given board and meet the fixture criteria that are specified.

> 
> When running the following:
> 
> ```
> sanitycheck -p disco_l475_iot1 -b -T tests/kernel/common/
> ```
> 
> and
> 
> ```
> sanitycheck -p disco_l475_iot1 --device-testing --hardware-map ../map.yaml -T tests/kernel/common/
> ```
> 
> the only difference I get in output artefacts are sanitycheck.csv, sanitycheck.log and sanitycheck.xml, which are only related to the pure execution (please correct me if I'm wrong).
> So, I don't really get what the `--build-only` would add in that context.

In some cases there will be no difference.  However if you do:
`sanitycheck -p disco_l475_iot1 -b -T tests/drivers/build_all/`

vs

`sanitycheck -p disco_l475_iot1 -b --device-testing --hardware-map ../map.yaml -T tests/drivers/build_all/`

You should see a difference.

**Comment by erwango** [comment]
>  In some cases there will be no difference. However if you do:
> sanitycheck -p disco_l475_iot1 -b -T tests/drivers/build_all/
> 
> vs
> 
> sanitycheck -p disco_l475_iot1 -b --device-testing --hardware-map ../map.yaml -T tests/drivers/build_all/
> 
> You should see a difference.

I see.
Anyway for average users, maybe a `--artefacts-only` would be less confusing than `--build-only`.

**Comment by nashif** [comment]
> Would we support `./scripts/sanitycheck --device-testing -p qemu_x86` ?

Hmm, no, --device-testing is mostly about the handler (DeviceHandler), flashing and connecting to the device and also managing multiple devices using the hardware map etc. 

generating artifacts for a specific board should be generic and orthogonal to the handler and how you interact with the device.

File: scripts/sanity_chk/sanitylib.py [hunk 1]
```diff
@@ -1602,6 +1602,11 @@ def check_build_or_run(self, build_only=False, enable_slow=False, device_testing
 
         self.build_only = not (not _build_only and runnable)
         self.run = not self.build_only
+
+        # we asked for build-only on the command line
+        if opt_build_only:
+            self.build_only = True
```
Comment (PerMac) [inline scripts/sanity_chk/sanitylib.py:1608]: I think that when this check is just after `self.run = not self.build_only` we could end up (at least in theory) with both `self.build_only` and `self.run` == `True`. Isn't this an issue?

Comment (PerMac) [reply scripts/sanity_chk/sanitylib.py:1608]: I think I got it now. This is actually fine and would happen if the test is perfectly able to be run on a platform (and doesn't have build_only set to true) but the user used option from the cli to only build. I am right now?

Comment (galak) [reply scripts/sanity_chk/sanitylib.py:1608]: Correct.  This trying to cleanup the use of build_only in several places having slightly different meanings.

**Review by nashif — CHANGES_REQUESTED** [review]
getting buildable/runnable tests for a certain device should be orthogonal to --device-testing.

**Comment by galak** [comment]
> getting buildable/runnable tests for a certain device should be orthogonal to --device-testing.

How flags (existing or new) should be passed to sanitycheck for this purpose?

**Comment by nashif** [comment]
> What flags (existing or new) should be passed to sanitycheck for this purpose?

something like that maybe:

```sanitycheck -p qemu_x86 --build-only --filter runnable|buildable```

**Comment by nashif** [comment]
I started looking at this, getting something like this with initial prototype..

```
i9:zephyr(sanitycheck/build_only_cleanup): sanitycheck  -p native_posix --filter=runnable  -v -c -y
Deleting output directory /home/nashif/Work/zephyrproject/zephyr/sanity-out
INFO    - JOBS: 20
INFO    - Building initial testcase list...
INFO    - 323 test configurations selected, 544 configurations discarded due to filters.
INFO    - Completed in 0 seconds
```

vs.

```
i9:zephyr(sanitycheck/build_only_cleanup): sanitycheck  -p native_posix   -v -c -y
Deleting output directory /home/nashif/Work/zephyrproject/zephyr/sanity-out
INFO    - JOBS: 20
INFO    - Building initial testcase list...
INFO    - 388 test configurations selected, 479 configurations discarded due to filters.
INFO    - Completed in 0 seconds
```

**Comment by nashif** [comment]
along the way cleaned up the build_only ambiguity and the logic behind it for real...

**Comment by galak** [comment]
Neat.  Was thinking about a `--filter` option as well recently with all the various changes we got going on.

**Comment by galak** [comment]
Closing in favor of #27236

**Pull request closed without merging**